crates/polars-utils/src/cache.rs

```diff
@@ -29,3 +38,197 @@ where
         }
     }
 }
+
+/// A fixed-size cache optimized for access speed. Does not implement LRU or use
+/// a full hash table due to cost, instead we assign two pseudorandom slots
+/// based on the hash of the key, and if both are full we evict the one that had
+/// the older last access.
+const MIN_FAST_FIXED_CACHE_SIZE: usize = 16;
+
+#[derive(Clone)]
+pub struct FastFixedCache<K, V> {
+    slots: Vec<CacheSlot<K, V>>,
+    access_ctr: Cell<u32>,
+    shift: u32,
+    hash_builder: RandomState,
+}
+
+impl<K: Hash + Eq, V> FastFixedCache<K, V> {
+    pub fn new(n: usize) -> Self {
+        let n = (n.max(MIN_FAST_FIXED_CACHE_SIZE)).next_power_of_two();
+        Self {
+            slots: zeroed_vec(n),
+            access_ctr: Cell::new(1),
+            shift: 64 - n.ilog2(),
+            hash_builder: RandomState::new(),
+        }
+    }
+
+    pub fn get<Q: Hash + Eq + ?Sized>(&self, key: &Q) -> Option<&V>
+    where
+        K: Borrow<Q>,
+    {
+        unsafe {
+            // SAFETY: slot_idx from raw_get is valid and occupied.
+            let slot_idx = self.raw_get(self.hash(key), key)?;
+            let slot = self.slots.get_unchecked(slot_idx);
+            Some(slot.value.assume_init_ref())
+        }
+    }
+
+    pub fn get_mut<Q: Hash + Eq + ?Sized>(&mut self, key: &Q) -> Option<&mut V>
+    where
+        K: Borrow<Q>,
+    {
+        unsafe {
+            // SAFETY: slot_idx from raw_get is valid and occupied.
+            let slot_idx = self.raw_get(self.hash(&key), key)?;
+            let slot = self.slots.get_unchecked_mut(slot_idx);
+            Some(slot.value.assume_init_mut())
+        }
+    }
+
+    pub fn insert(&mut self, key: K, value: V) -> &mut V {
+        unsafe { self.raw_insert(self.hash(&key), key, value) }
+    }
+
+    pub fn get_or_insert_with<Q, F>(&mut self, key: &Q, f: F) -> &mut V
+    where
+        K: Borrow<Q>,
+        Q: Hash + Eq + ToOwned<Owned = K> + ?Sized,
+        F: FnOnce(&K) -> V,
+    {
+        unsafe {
+            let h = self.hash(key);
+            if let Some(slot_idx) = self.raw_get(self.hash(&key), key) {
+                let slot = self.slots.get_unchecked_mut(slot_idx);
+                return slot.value.assume_init_mut();
+            }
+
+            let key = key.to_owned();
```

Comment:
Do we need an owned version? Wouldn't a map of borrowed strings be valuable?

Reply:
Sure, and you can. `ToOwned` is implemented on `str`, not on `&str`. So `&str` becomes `String` with `ToOwned`, but `&&str` becomes `&str`.

Reply:
Today I learned about `to_owned` for `&&str`. Nice generalization.